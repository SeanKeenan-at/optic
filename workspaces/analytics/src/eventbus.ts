import { EventEmitter } from 'events';
import { v4 as uuidv4 } from 'uuid';

import {
  ClientContext,
  TrackingEventBase,
} from './interfaces/TrackingEventBase';

export function newAnalyticsEventBus(
  getContext: (batchId: string) => Promise<ClientContext>
): AnalyticsEventBus {
  const eventEmitter = new EventEmitter();
  return {
    emit: async (...events: TrackingEventBase<any>[]) => {
      const batchId = uuidv4();
      const context = await getContext(batchId);
      events.forEach((event) => {
        eventEmitter.emit('event', { event, context });
      });
    },
    eventEmitter,
    listen: (callback: ListenCallbackType): void => {
      eventEmitter.on('event', callback);
    },
  };
}

type ListenCallbackType = (e: {
  event: TrackingEventBase<any>;
  context: ClientContext;
}) => void;

export interface AnalyticsEventBus {
  eventEmitter: EventEmitter;
  emit: (...events: TrackingEventBase<any>[]) => void;
  listen: (callback: ListenCallbackType) => void;
}
