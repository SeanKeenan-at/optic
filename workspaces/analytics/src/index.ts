export * from './cliEvents';
export * from './eventbus';
export * from './uiEvents';
export * from './interfaces/TrackingEventBase';
