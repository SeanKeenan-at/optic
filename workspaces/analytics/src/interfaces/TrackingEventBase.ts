export interface TrackingEventBase<T> {
  type: string;
  data: T;
}

export interface CliClientContext {
  clientId: string;
  platform: string;
  arch: string;
  release: string;
  apiName: string;
  clientSessionInstanceId: string;
  clientTimestamp: string;
  clientAgent: string;
  source: string;
}
