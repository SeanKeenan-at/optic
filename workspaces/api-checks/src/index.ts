export * from './runner';
export * from './spectacle';
export * as checks from './checks';
export * as reports from './reports';
