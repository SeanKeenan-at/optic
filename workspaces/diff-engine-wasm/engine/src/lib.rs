#![allow(dead_code, unused_imports, unused_variables)]

use chrono::Utc;
use uuid::Uuid;
use wasm_bindgen::prelude::*;

use optic_diff_engine::{
  CommandContext, HttpInteraction, InteractionDiffResult, SpecCommand, SpecEvent, SpecProjection,
};

#[wasm_bindgen(start)]
pub fn init() {
  wasm_logger::init(wasm_logger::Config::default());
  std::panic::set_hook(Box::new(console_error_panic_hook::hook));
}

#[wasm_bindgen]
pub fn spec_from_events(spec_json: String) -> Result<WasmSpecProjection, JsValue> {
  let spec_events: Vec<SpecEvent> = serde_json::from_str(&spec_json).unwrap();
  let spec_projection = SpecProjection::from(spec_events);

  Ok(WasmSpecProjection::from(spec_projection))
}

#[wasm_bindgen]
pub fn diff_interaction(
  interaction_json: String,
  spec: &WasmSpecProjection,
) -> Result<String, JsValue> {
  let interaction: HttpInteraction = serde_json::from_str(&interaction_json).unwrap();

  let results: Vec<ResultContainer<InteractionDiffResult>> = spec
    .diff_interaction(interaction)
    .into_iter()
    .map(|result| result.into())
    .collect();

  Ok(serde_json::to_string(&results).unwrap())
}

#[wasm_bindgen]
pub struct WasmSpecProjection {
  projection: SpecProjection,
}

impl WasmSpecProjection {
  pub fn diff_interaction(&self, interaction: HttpInteraction) -> Vec<InteractionDiffResult> {
    optic_diff_engine::diff_interaction(&self.projection, interaction)
  }
}

impl From<SpecProjection> for WasmSpecProjection {
  fn from(projection: SpecProjection) -> Self {
    Self { projection }
  }
}

#[derive(Debug, serde::Deserialize, serde::Serialize)]
struct ResultContainer<T>(T, String);

impl From<InteractionDiffResult> for ResultContainer<InteractionDiffResult> {
  fn from(result: InteractionDiffResult) -> Self {
    let fingerprint = result.fingerprint();
    Self(result, fingerprint)
  }
}

// Commit batch
// ------------

#[wasm_bindgen]
pub fn append_batch_to_spec(
  spec: WasmSpecProjection,
  commands: String,
  commit_message: String,
) -> Result<String, JsValue> {
  let commands: Vec<SpecCommand> = serde_json::from_str(&commands).unwrap();
  let batch_id = Uuid::new_v4().to_hyphenated().to_string();
  let batch_command_context = CommandContext::new(
    batch_id.clone(),
    String::from("diff-engine-wasm-user"),
    String::from("diff-engine-wasm-session"),
    Utc::now(),
  );

  let mut batch =
    optic_diff_engine::append_batch_to_spec(spec.projection, commit_message, batch_command_context);

  for command in commands {
    batch
      .with_command(command)
      .map_err(|err| format!("command could not be applied: {:?}", err))?;
  }

  let new_events = batch.commit();

  serde_json::to_string(&new_events)
    .map_err(|err| JsValue::from(format!("new events could not be serialized: {:?}", err)))
}
