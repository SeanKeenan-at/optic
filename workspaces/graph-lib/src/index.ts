import * as endpoints from './endpoints-graph';
import * as shapes from './shapes-graph';

export { endpoints, shapes };
