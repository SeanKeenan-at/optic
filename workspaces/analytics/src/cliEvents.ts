import { CheckAssertionsResult } from './interfaces/ApiCheck';
import { TrackingEventBase } from './interfaces/TrackingEventBase';

export enum EventsEnum {
  // Onboarding
  UserLoggedInFromCLI = 'UserLoggedInFromCLI',
  ApiInitializedInProject = 'ApiInitializedInProject',
  ApiCheckCompleted = 'ApiCheckCompleted',

  // running tasks
  StartedTaskWithLocalCli = 'StartedTaskWithLocalCli',
  ExitedTaskWithLocalCli = 'ExitedTaskWithLocalCli',

  StatusRun = 'StatusRun',
  LiveTrafficIngestedWithLocalCli = 'LiveTrafficIngestedWithLocalCli',
}

type UserLoggedInFromCLIProps = {
  userId: string;
};

export const UserLoggedInFromCLI = (
  data: UserLoggedInFromCLIProps
): TrackingEventBase<UserLoggedInFromCLIProps> => ({
  type: EventsEnum.UserLoggedInFromCLI,
  data,
});

type ApiInitializedInProjectProps = {
  apiName: string;
  source: string;
  cwd: string;
};

export const ApiInitializedInProject = (
  data: ApiInitializedInProjectProps
): TrackingEventBase<ApiInitializedInProjectProps> => ({
  type: EventsEnum.ApiInitializedInProject,
  data,
});

export const ApiCheckCompleted = (
  data: CheckAssertionsResult
): TrackingEventBase<CheckAssertionsResult> => ({
  type: EventsEnum.ApiCheckCompleted,
  data,
});

type StartedTaskWithLocalCliProps = {
  captureId: string;
  cwd: string;
  inputs: {
    task: string;
  } & Record<string, string>;
};

export const StartedTaskWithLocalCli = (
  data: StartedTaskWithLocalCliProps
): TrackingEventBase<StartedTaskWithLocalCliProps> => ({
  type: EventsEnum.StartedTaskWithLocalCli,
  data,
});

type ExitedTaskWithLocalCliProps = {
  interactionCount: number;
  captureId: string;
  inputs: {
    task: string;
  } & Record<string, string>;
};

export const ExitedTaskWithLocalCli = (
  data: ExitedTaskWithLocalCliProps
): TrackingEventBase<ExitedTaskWithLocalCliProps> => ({
  type: EventsEnum.StartedTaskWithLocalCli,
  data,
});

type StatusRunProps = {
  captureId: string;
  diffCount: number;
  undocumentedCount: number;
  timeMs: number;
};

export const StatusRun = (
  data: StatusRunProps
): TrackingEventBase<StatusRunProps> => ({
  type: EventsEnum.StatusRun,
  data,
});

type LiveTrafficIngestedWithLocalCliProps = {
  captureId: string;
  interactionCount: number;
};

export const LiveTrafficIngestedWithLocalCli = (
  data: LiveTrafficIngestedWithLocalCliProps
): TrackingEventBase<LiveTrafficIngestedWithLocalCliProps> => ({
  type: EventsEnum.StatusRun,
  data,
});
