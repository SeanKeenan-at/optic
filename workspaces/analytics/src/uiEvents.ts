import { TrackingEventBase } from './interfaces/TrackingEventBase';

type AnalyticsDispatcher = (event: TrackingEventBase<any>) => void;

export class OpticUIEvents {
  constructor(private dispatch: AnalyticsDispatcher) {}

  // in-use
  // documentation page
  documentationPageLoaded() {
    this.dispatch({
      type: 'documentation_page_loaded',
      data: {},
    });
  }

  documentationListPageLoaded() {
    this.dispatch({
      type: 'documentation_list_page_loaded',
      data: {},
    });
  }

  userSavedDocChanges(
    deletedEndpointCount: number,
    contributionChangesCount: number,
    specId: string
  ) {
    this.dispatch({
      type: 'user_saved_documentation_changes',
      data: { deletedEndpointCount, contributionChangesCount, specId },
    });
  }

  userStartedSharing() {
    this.dispatch({
      type: 'user_started_sharing',
      data: {},
    });
  }

  userPickedShareTarget(shareWith: string) {
    this.dispatch({
      type: 'user_picked_share_target',
      data: { with: shareWith },
    });
  }

  userShared(shareWith: string, publicSpecId: string) {
    this.dispatch({
      type: 'user_saved_documentation_changes',
      data: { with: shareWith, publicSpecId },
    });
  }

  userDeletedEndpoint() {
    this.dispatch({
      type: 'user_deleted_endpoint',
      data: {},
    });
  }

  // diff page
  userResetDiff(currentApproved: number, totalDiffs: number) {
    this.dispatch({
      type: 'user_reset_diff',
      data: { currentApproved, totalDiffs },
    });
  }

  reviewPageLoaded(
    diffs: number,
    undocumentedUrls: number,
    diffEngineDurationMilliseconds: number,
    endpointsInSpec: number
  ) {
    this.dispatch({
      type: 'review_page_loaded',
      data: {
        undocumentedUrls,
        diffEngineDurationMilliseconds,
        endpointsInSpec,
      },
    });
  }

  userApprovedAll(shapeDiffs: number, newBodyDiffs: number) {
    this.dispatch({
      type: 'user_approved_all',
      data: { shapeDiffs, newBodyDiffs },
    });
  }

  userDocumentedEndpoint(isBulkMode: boolean) {
    this.dispatch({
      type: 'user_documented_endpoint',
      data: { isBulkMode },
    });
  }

  userDiscardedEndpoint() {
    this.dispatch({
      type: 'user_discarded_endpoint',
      data: {},
    });
  }

  userChoseACaptureMethod(captureSource: string) {
    this.dispatch({
      type: 'user_chose_capture_method',
      data: { captureSource },
    });
  }

  userSavedChanges(
    undocumentedUrls: number,
    endpointsChangedCount: number,
    specId: string
  ) {
    this.dispatch({
      type: 'user_saved_changes',
      data: {
        undocumentedUrls,
        endpointsChangedCount,
        specId,
      },
    });
  }

  // history page
  historyPageLoaded() {
    this.dispatch({
      type: 'history_page_loaded',
      data: {},
    });
  }

  resetToCommit(numberOfCommitsReset: number) {
    this.dispatch({
      type: 'reset_to_commit',
      data: {
        numberOfCommitsReset,
      },
    });
  }

  // Temporary
  // 2021/07/13- can remove 2021/10/13
  queryParameterDocumented(numberOfQueryParameterBodies: number) {
    this.dispatch({
      type: 'query_parameter_documented',
      data: {
        numberOfQueryParameterBodies,
      },
    });
  }

  // proposed

  // deprecated
}
