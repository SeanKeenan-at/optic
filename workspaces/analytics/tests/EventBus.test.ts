import { newAnalyticsEventBus } from '../src/eventbus';
import { ClientContext } from '../src/interfaces/TrackingEventBase';
const { ApiCreated } = require('../src/events/onboarding');

const getContext: (batchId: string) => Promise<ClientContext> = async () => {
  return {
    clientId: 'testing',
    platform: 'furby',
    arch: 'mips',
    release: '1998',
    apiName: 'testing',
    clientSessionInstanceId: 'testing',
    clientTimestamp: 'testing',
    clientAgent: 'testing',
  };
};

test('Event Bus can listen for analytics events', async (done) => {
  const bus = newAnalyticsEventBus(getContext);

  const promise = new Promise((resolve) => {
    bus.listen((e) => {
      expect(e).toMatchSnapshot();
      resolve();
    });
  }).then(done);

  bus.emit(ApiCreated.withProps({ apiName: 'Hello World API' }));
});
