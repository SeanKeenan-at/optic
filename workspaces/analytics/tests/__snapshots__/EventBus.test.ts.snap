// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event Bus can listen for analytics events 1`] = `
Object {
  "context": Object {
    "clientAgent": "testing",
    "clientId": "testing",
    "platform": "furby",
    "arch": "mips",
    "release": "1998",
    "apiName": "testing",
    "clientSessionInstanceId": "testing",
    "clientTimestamp": "testing",
  },
  "data": Object {
    "apiName": "Hello World API",
  },
  "type": "ApiCreated",
}
`;
